paparazzi/paparazzi-gradle-plugin/src/main/java/app/cash/paparazzi/gradle/PaparazziPlugin.kt

```diff
@@ -163,6 +163,14 @@ class PaparazziPlugin : Plugin<Project> {
         test.outputs.dir(snapshotOutputDir)
 
         val paparazziProperties = project.properties.filterKeys { it.startsWith("app.cash.paparazzi") }
+        test.systemProperties.putAll(paparazziProperties)
```

Comment:
why did this have to move?  looks irrelevant to the change?

Reply:
Because it's irrelevant to the doFirst hack (didn't look at history but possible that it was added here because doFirst was already setting properties). Now that I had full understanding I could confidently move it outside the doFirst block. It has the same problem as the other two, if it's set AFTER up to date check, it breaks Gradle. see https://github.com/cashapp/paparazzi/pull/690/commits/8ab5ee9f38b288bbdb41324d08b913cf7639fd9b
I can't prove this, because I don't know what those properties can be, but I'm almost certain they would come from `-Papp.cash.paparazzi.foo` if you point me to how/where/why this is used, I can add another test proving it was broken.

Reply:
Sorry, still not following.  Rewording my question: does moving this line have any effect on fixing the "make sure record/verify actually affect test tasks" original premise that linked to the comment?   it doesn't appear so, so should probably be undone?

> because I don't know what those properties can be

I think the only example at the moment is the property that enables turning off verbose logging?  https://github.com/cashapp/paparazzi/blob/master/paparazzi/paparazzi-gradle-plugin/src/test/projects/flag-debug-linked-objects-on/gradle.properties#L2

And that wouldn't/shouldn't affect test task cacheability?

Reply:
Yes, the original issue @alashow reported was that record/verify was not re-executing if test was executed before. Actually he only reported record, I read the code and added verify because I saw it was broken the same way. I also added properties to the scope because it was broken the same way.

Everything that modifies a tasks behaviour must cause the task to re-execute. If it doesn't, it's going to cause incorrect builds and/or confuse users, because that's just how Gradle tasks are expected to work. This is especially true for something that changes logging, because imagine I run:
```
gradlew recordPaparazziDebug
```
and then I realize I want to see the logs for it, so I run:
```
gradlew recordPaparazziDebug -Papp.cash.paparazzi.debug.linked.objects=true
```
and I get `> :testDebugUnitTest UP-TO-DATE`, that's a big wtf moment. Because I changed a build parameter! I won't change dependencies, because I want to see the current version's logs. I won't change my code, because I want to see the current code's behaviour. I won't change the tests, because I want to see my current test behaviour. I **will** add the property, because the change is that before I didn't want to see logs, and now I want to see them.

Add on top of this if the user hopefully has caching enabled and because the task executed unexpectedly UP-TO-DATE, now the next reasonable thing is to do:
```
gradlew clean recordPaparazziDebug -Papp.cash.paparazzi.debug.linked.objects=true
```
(which one should never do, btw.)
...and now I get `> :testDebugUnitTest FROM-CACHE`, there's just no winning trying to execute my tests with logs. User has to hack around with even more advanced Gradle options or change their code for no reason to get their logs from Paparazzi.

With this PR neither of this will happen it will "just work", this PR essentially changes the cache key, which is used for both build cache and up to date checks. If the user removed the flag, the old cache key will be used and tests will execute from cache.

So, yes, even for a logging altering behaviour, cache key should change and affected tasks re-execute.

Note that enabling logs will affect the console output, but will also affect the generated report (Standard Output) tab. So it seriously changes the `@OutputDirectory` contents, it is therefore an `@Input` to the task. `systemProperties` is automatically an input, so this works.

Note that changing a value in `gradle.properties` and `-P` and `ORG_GRADLE_PROJECT_prop` env var are all almost equivalent. They all show up in the same way in `Project.properties`, just with different precedence and override order.

Reply:
Sorry for the delay in getting back to this!

Re: how Gradle works, I understand the caching bit, but I think the part that I was missing (and now understand) is this: 

> With this PR neither of this will happen it will "just work", this PR essentially changes the cache key, which is used for both build cache and up to date checks. If the user removed the flag, the old cache key will be used and tests will execute from cache.

TL;DR: setting the system properties in the task action is too late in terms of having those properties considered in calculation of the cache key!  and I agree, changing the logging property should result in showing/hiding the logs which would require a re-run of the task.